### ISSUE TITLE
Delay running slot lambda until after the `render_in` block

### ISSUE THREAD

**Issue by TastyPi** [ViewComponent/view_component#1527]
## Feature request

It would be nice if slot lambdas were not called until after the block passed to `render_in` is called.

### Motivation

Sometimes I need to change the behaviour of a slot depending on the existence of another slot. A good example of this is a form field component that supports displaying an error message:

```ruby
class FieldComponent < ViewComponent::Base
  renders_one input, ->(type: "text") { text_field_tag name, value, class: ("error" if error?), type: }
  renders_one error
end
```

If this component is rendered like this:

```ruby
render FieldComponent.new do |field|
  field.with_error { "Error message" }
  field.with_input
end
```

then the `error` class is added as expected, however the following does not include the `error` class as `error?` returns `false`:

```ruby
render FieldComponent.new do |field|
  field.with_input
  field.with_error { "Error message" }
end
```

Unfortunately in this scenario the code reads better if the `with_input` comes first. If the lambda was called after the block finished this would not be an issue. I believe this would also be the expected behaviour for most people as you would expect the order of calling `with_` methods for `renders_one` slots to not matter.

**Comment by TastyPi** [comment]
A hacky proof-of-concept workaround is to have the lambda return a component:

```ruby
class DelayedRenderComponent < ViewComponent::Base
  def initialize(&block)
    @block = block
  end

  def call
    block.call
  end
end

class FieldComponent < ViewComponent::Base
  renders_one input, ->(type: "text") do
    DelayedComponent.new { text_field_tag name, value, class: ("error" if error?), type: }
  end
  renders_one error
end
```

Looks like it would be fairly easy to update the implementation here to support it
 https://github.com/ViewComponent/view_component/blob/88cec826bb43eb070f66acd0b243b22f29f7e178/lib/view_component/slotable_v2.rb#L354-L374

Instead of calling the function and playing with the value it should be possible to create a basic class which calls the function in its `render_in` method, then set that to the `slot.__vc_component_instance`.

I don't have time to dig further now, but I think I'll have time in the next couple of weeks to look at implementing it.

**Comment by joelhawksley** [comment]
@TastyPi thanks for writing this up. I'm cautious changing this behavior as I could imagine others depending on the existing behavior.

Looking at your specific example, perhaps you could set the error as part of the `input` slot instead of having separate slots?

**Comment by TastyPi** [comment]
So I want to be able to use the component something like this:

```erb
<%= render FieldComponent.new do |field| %>
  <% field.input(...) %>
  <% field.error do %>
    <ul>
      <% errors.each do |error| %>
        <li><%= error %>
      <% end %>
    </ul>
  <% end %>
```

There are definitely ways I could make this work other than forcing the `input` to be rendered lazily, e.g. setting the `error` slot first, specifying `error: true` somewhere it can be used in the `input` slot, or moving the CSS class to a parent of the `input` element and have the CSS work with that structure.

My bigger problem is the behaviour is inconsistent and undocumented. For example, I would expect slots like these to always behave the same way:

```erb
renders_one :eager, ->() { render MyComponent.new(...) }
renders_one :lazy, ->() { MyComponent.new(...) }
```

However, if the slot depends on another component they behave differently.

I understand not wanting to introduce breaking changes, but with version 3 coming up I think it would be nice to make this behaviour consistent (and documented) with said major version bump. There are two options I believe make sense - render the slot when the slot setter is called (eager), or render the slot when it is first used while rendering the component (lazy). Obviously my use-case would work better with the lazy behaviour, but I'd also be fine with eager and picking a different workaround, I'd much rather the behaviour be consistent. When I was first building the component I found it confusing when `error?` would not always work, if the behaviour of when a slot is rendered was predictable and documented it would have saved a lot of confusion.

There's also this very contrived example where the behaviour differs but in a way that favours eager rendering:

```rb
class ExampleComponent < ViewComponent::Base
  renders_one :eager, ->(&block) { render ExampleSlotComponent.new(block) }
  renders_one :lazy, ->(&block) { ExampleSlotComponent.new(block) }
end

class ExampleSlotComponent < ViewComponent::Base
  def initialize(block)
    @block = block
  end

  def call
    @block.call
  end
end 
```

```erb
<%= render ExampleComponent.new do |example| %>
  <% x = 1 %>
  <% example.eager { x } %><%# renders 1 %>
  <% example.lazy { x } %><%# renders 2 %>
  <% x = 2 %>
  <% example.eager { x } %><%# renders 2 %>
  <% example.lazy { x } %><%# renders 2 %>
<% end %>
```

Personally, I kind of think eager rendering is more intuitive but lazy rendering is more useful, so it comes down to what's more important for `view_component`.

**Comment by TastyPi** [comment]
> I can see why it might be misleading, but the behavior here is consistent, which is that the lambda is always eagerly evaluated.

Fair enough, I hadn't thought of it from that angle but I do agree that makes sense.

> I think we could document the difference between returning an object that responds to render_in vs returning a string if it's not already, though.

This would be the main I want,

**Comment by TastyPi** [comment]
> I can see why it might be misleading, but the behavior here is consistent, which is that the lambda is always eagerly evaluated.

Fair enough, I hadn't thought of it from that angle, but I can agree that it makes sense. I do still think that returning a component or returning the string from rendering a component should result in the same thing, but that's just an opinion that would have been immediately changed if I was corrected by the documentation. Speaking of...

> I think we could document the difference between returning an object that responds to render_in vs returning a string if it's not already, though.

I couldn't find it mentioned anywhere, I would expect to find it here https://viewcomponent.org/guide/slots.html#lambda-slots.

If all that changes from a result of this issue is a documentation update (and maybe a test to ensure the behaviour doesn't change) I'd be happy with that.

(BTW I don't need more solutions to the specific example I gave, I have plenty of options and the component it was originally based on is very different now. Your first solution with the `set_error` is quite nice though.)
